### PR TITLE
Align KTO with DPO: Align ref_model initialization

### PR DESCRIPTION
**Pull request by albertvillanova**
Align KTO with DPO: Align ref_model initialization.

This PR refactors how the reference model (`ref_model`) is handled in the `KTOTrainer` class, making its initialization more robust, clarifying its usage, and improving compatibility with PEFT models. The changes ensure that `ref_model` is either explicitly provided or automatically derived from the initial model state, with special handling for PEFT adapters.

Part of:
- #4786

### Changes

Reference model initialization and handling:

* The `ref_model` parameter in `KTOTrainer` is now expected to be either a `PreTrainedModel` or `None`, removing support for passing a string path directly. If `ref_model` is not provided, the trainer will automatically initialize it from the initial policy or, if using a PEFT model, set it to `None` since the adapter can be disabled to revert to the initial model.
* Added a check to prevent `model` and `ref_model` from being the same object, raising a clear error if this occurs and guiding the user to omit `ref_model` for automatic initialization.
* Updated the docstring for `ref_model` in both `KTOTrainer` and `DPOTrainer` to clarify its purpose and usage, including the new initialization logic.

Internal utilities and imports:

* Added import of `get_config_model_id` and updated imports to support the new reference model initialization logic.

---

> [!NOTE]
> **Medium Risk**
> Changes how `KTOTrainer` derives/loads the reference model, which can affect training dynamics and memory usage, especially for non-PEFT runs and distributed setups. While the logic mirrors `DPOTrainer`, misconfiguration could surface as runtime errors or different ref-policy behavior.
> 
> **Overview**
> Aligns `KTOTrainer` reference-policy handling with `DPOTrainer`: `ref_model` is now *optional* and, when omitted, the trainer reconstructs a reference model from `model.config` via `get_config_model_id` (or uses `None` for PEFT so adapters can be disabled to recover the initial policy).
> 
> Removes support for passing `ref_model` as a string/path and reorders setup so dropout disabling also applies to the newly-initialized reference model; also tightens validation by rejecting `ref_model is model` with updated guidance. Updates the `DPOTrainer` docstring type annotation for `ref_model` for consistency.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 7df5ac11599a31e08ff11416596cfafeb696fe08. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/experimental/kto/kto_trainer.py [hunk 1]
```diff
@@ -252,10 +253,12 @@ class KTOTrainer(_BaseTrainer):
     Args:
         model ([`~transformers.PreTrainedModel`]):
             The model to train, preferably an [`~transformers.AutoModelForSequenceClassification`].
-        ref_model ([`~transformers.PreTrainedModel`]):
-            Hugging Face transformer model with a casual language modelling head. Used for implicit reward computation
-            and loss. If no reference model is provided, the trainer will create a reference model with the same
-            architecture as the model to be optimized.
+        ref_model ([`~transformers.PreTrainedModel`], *optional*):
```
Comment (qgallouedec) [inline trl/experimental/kto/kto_trainer.py:256]: related https://github.com/huggingface/trl/pull/5536

Comment (qgallouedec) [reply trl/experimental/kto/kto_trainer.py:256]: ah sorry, already on this pr

**Review by qgallouedec — APPROVED** [review]

File: trl/experimental/kto/kto_trainer.py [hunk 2]
```diff
@@ -679,6 +659,28 @@ def make_inputs_require_grad(module, input, output):
...
+                # If PEFT is used, the reference model is not needed since the adapter can be disabled to revert to the
+                # initial model.
+                self.ref_model = None
+            else:
+                ref_model_init_kwargs = args.model_init_kwargs or {}
+                # Distributed training requires device_map=None ("auto" fails)
+                if self.args.distributed_state.distributed_type in ["MULTI_GPU", "DEEPSPEED"]:
+                    ref_model_init_kwargs["device_map"] = None
+                ref_model_path = get_config_model_id(self.model.config)
+                self.ref_model = create_model_from_path(ref_model_path, **ref_model_init_kwargs)
```
[the remark this answers is not in the document]
Comment (albertvillanova) [reply trl/experimental/kto/kto_trainer.py:674]: I agree, but this is how DPO currently does. I will address in a separate PR the alignment of this specific case.
